include/deal.II/base/array_view.h

```diff
@@ -39,6 +39,142 @@ class Table;
 template <typename number>
 class LAPACKFullMatrix;
 
+/**
+ * Base class of @p ArrayView which allows strided access into the view.
+ * This is particularly useful when you want to access only one lane of a
+ * VectorizedArray.
+ */
+template <typename ElementType, std::size_t stride = 1>
+class StridedArrayView
+{
+public:
+  /**
+   * An alias that denotes the "value_type" of this container-like class,
+   * i.e., the type of the element it "stores" or points to.
+   */
+  using value_type = ElementType;
+
+  /**
+   * Constructor.
+   *
+   * @param[in] starting_element A pointer to the first element of the array
+   * this object should represent.
+   * @param[in] n_elements The length (in elements) of the chunk of memory
+   * this object should represent.
+   *
+   * @note The object that is constructed from these arguments has no
+   * knowledge how large the object into which it points really is. As a
+   * consequence, whenever you call ArrayView::operator[], the array view can
+   * check that the given index is within the range of the view, but it can't
+   * check that the view is indeed a subset of the valid range of elements of
+   * the underlying object that allocated that range. In other words, you need
+   * to ensure that the range of the view specified by the two arguments to
+   * this constructor is in fact a subset of the elements of the array into
+   * which it points. The appropriate way to do this is to use the
+   * make_array_view() functions.
+   */
+  StridedArrayView(value_type *starting_element, const std::size_t n_elements);
+
+  /**
+   * Return the size (in elements) of the view of memory this object
+   * represents.
+   */
+  std::size_t
+  size() const;
+
+  /**
+   * Return a bool whether the array view is empty.
+   */
+  bool
+  empty() const;
+
+  /**
+   * Return a pointer to the underlying array serving as element storage.
+   * In case the container is empty a nullptr is returned.
+   */
+  DEAL_II_HOST_DEVICE value_type *
```

Comment:
Does the error have something to do with this macro? Do I have to template `StridedArrayView` over `MemorySpaceType` to avoid this?
```
 /home/runner/work/dealii/dealii/include/deal.II/base/array_view.h:141:40: error: __host__ function 'data' cannot overload __host__ __device__ function 'data'
StridedArrayView<ElementType, stride>::data() const noexcept
                                       ^
/home/runner/work/dealii/dealii/include/deal.II/base/array_view.h:96:3: note: previous declaration is here
  data() const noexcept;
```

Reply:
Can somebody with more knowledge about memory spaces give me some insight here? I cannot reproduce this locally.

Reply:
You also need to annotate the definition of the function with `DEAL_II_HOST_DEVICE` (if this function is indeed supposed to be device-callable).

Reply:
Thanks @masterleinad ! Updated